File: LeetCode/1639. Number of Ways to Form a Target String Given a Dictionary/solution.py
# from functools import lru_cache
# class Solution:
#     def numWays(self, words: List[str], target: str) -> int:
#         mp = defaultdict(int)
#         MOD = 10**9+7
#         n, m = len(target), len(words[0])
#         for word in words:
#             for i, w in enumerate(word):
#                 mp[w,i] += 1

#         @lru_cache(None)
#         def dfs(t,k):
#             if t == n:
#                 return 1
#             if k == m:
#                 return 0
#             return sum([mp[target[t],j]*dfs(t+1,j+1) for j in range(k,m) if mp[target[t],j] > 0])%MOD
#         return dfs(0,0)



from functools import lru_cache
class Solution:
    def numWays(self, words: List[str], target: str) -> int:
        mp = defaultdict(int)
        MOD = 10**9+7
        n, m = len(target), len(words[0])
        for word in words:
            for i, w in enumerate(word):
                mp[w,i] += 1

        @lru_cache(None)
        def dfs(t,k):
            if t == n:
                return 1
            if k == m:
                return 0
            res = dfs(t,k+1)
            if mp[target[t],k] > 0:
                res += mp[target[t],k]*dfs(t+1,k+1)
            return res%MOD
            # return (dfs(t,k+1) + mp[target[t],k]*dfs(t+1,k+1))%MOD
            # return sum([mp[target[t],j]*dfs(t+1,j+1) for j in range(k,m) if mp[target[t],j] > 0])%MOD
        return dfs(0,0)


class Solution:
    def numWays(self, words: List[str], target: str) -> int:
        mp = defaultdict(int)
        MOD = 10**9+7
        n, m = len(target), len(words[0])
        for word in words:
            for i, w in enumerate(word):
                mp[w,i] += 1

        dp = [0]*(n+1)
        dp[-1] = 1
        for k in range(m)[::-1]:
            for t in range(n):
                dp[t] += dp[t+1]*mp[target[t],k]
        return dp[0]%MOD

File: LeetCode/1639. Number of Ways to Form a Target String Given a Dictionary/solution.cpp
using ll = long long;

class Solution {
public:
    int numWays(vector<string>& words, string target) {
        int sz = words.size(), m = words[0].length(), n = target.length();
        int MOD = 1e9+7;
        vector<vector<int>> cc(m,vector<int>(26,0));
        for (int i = 0; i < sz; i++){
            for (int j = 0; j < m; j++)
                cc[j][words[i][j]-'a']++;
        }
        vector<ll> dp(m,1);
        
        for (int i = 0; i < n; i++){
            ll pre = i == 0 ? 1 : 0;
            for (int j = 0; j < m; j++){
                int tmp = dp[j];
                dp[j] = j == 0 ? 0 : dp[j-1];
                dp[j] += pre*cc[j][target[i]-'a'];
                dp[j] %= MOD;
                pre = tmp;
            }
        }
        return dp.back();
    }
};